unblob/handlers/filesystem/jffs2.py

```diff
@@ -0,0 +1,150 @@
+import io
+from enum import IntEnum
+from typing import List, Optional
+
+from structlog import get_logger
+
+from unblob.file_utils import Endian, convert_int16, read_until_past, round_up
+
+from ...models import StructHandler, ValidChunk
+
+logger = get_logger()
+
+
+BLOCK_ALIGNMENT = 4
+
+# Compatibility flags.
+JFFS2_COMPAT_MASK = 0xC000
+JFFS2_NODE_ACCURATE = 0x2000
+JFFS2_FEATURE_INCOMPAT = 0xC000
+JFFS2_FEATURE_ROCOMPAT = 0x8000
+JFFS2_FEATURE_RWCOMPAT_COPY = 0x4000
+JFFS2_FEATURE_RWCOMPAT_DELETE = 0x0000
+
+JFFS2_MAGICS = [0x1985, 0x8519, 0x1984, 0x8419]
+
+
+class JFFS2_NODETYPE(IntEnum):
+    DIRENT = JFFS2_FEATURE_INCOMPAT | JFFS2_NODE_ACCURATE | 1
+    INODE = JFFS2_FEATURE_INCOMPAT | JFFS2_NODE_ACCURATE | 2
+    CLEANMARKER = JFFS2_FEATURE_RWCOMPAT_DELETE | JFFS2_NODE_ACCURATE | 3
+    PADDING = JFFS2_FEATURE_RWCOMPAT_DELETE | JFFS2_NODE_ACCURATE | 4
+    SUMMARY = JFFS2_FEATURE_RWCOMPAT_DELETE | JFFS2_NODE_ACCURATE | 6
+    XATTR = JFFS2_FEATURE_INCOMPAT | JFFS2_NODE_ACCURATE | 8
+    XREF = JFFS2_FEATURE_INCOMPAT | JFFS2_NODE_ACCURATE | 9
+
+
+class _JFFS2Base(StructHandler):
+    C_DEFINITIONS = r"""
+        typedef struct jffs2_unknown_node
+        {
+            /* All start like this */
+            uint16 magic;
+            uint16 nodetype;
+            uint32 totlen; /* So we can skip over nodes we don't grok */
+            uint32 hdr_crc;
+        } jffs2_unknown_node_t;
+    """
+
+    HEADER_STRUCT = "jffs2_unknown_node_t"
+
+    BIG_ENDIAN_MAGIC = 0x19_85
+
+    def guess_endian(self, file: io.BufferedIOBase) -> Endian:
+        magic = convert_int16(file.read(2), Endian.BIG)
```

Comment:
what happens if we are at the end of the file? I guess we'd raise from `convert_int16` and throw away this chunk's processing. This shouldn't be an issue, just wanted to think through it :)

Interestingly `file_utils.get_endian` has different behavior, as it would seek back more when it is near the end: https://github.com/IoT-Inspector/unblob/pull/157

Reply:
We call `guess_endian` only once on the beginning of our match. Given that our yara rule match on exactly 4 bytes, `convert_int16` will never throw an `EOFError`.